### ISSUE TITLE
Dynamic commands based on user input

### ISSUE THREAD

**Issue by mminer** [asabaylus/react-command-palette#362]
**Is your feature request related to a problem? Please describe.**
Currently I'm unable to generate commands dynamically based on the user's query. For example, when the user types "my query", I would like to include commands `Search for "my query"` and `Create file named "myQuery.txt"`.

**Describe the solution you'd like**
Ideally we'd have a mechanism to access the `<input>`'s value. A solution that comes to mind is allowing the `commands` prop to also accept a `(query: string) => object[]` callback, passing the user's query as an argument.

**Describe alternatives you've considered**
Another solution could be exposing an `inputRef` prop which gets forwarded the `<input>` element, thus allowing the parent component to access the query directly via `inputRef.current.value`. I dislike this approach as it exposes implementation details about the component. The input field will probably always be a bona fide `<input>`, but still, providing the query via a callback seems superior.

I could probably use `findDOMNode` to get a reference to the `<input>` and regenerate the command array when the query changes, but obviously this is hacky and sad.

Happy to work on this. Or if this is possible already and I'm simply missing it, would love to know it.

**Comment by asabaylus** [comment]
Feel free to open a PR, if not I’ll see about adding it shortly

**Comment by asabaylus** [comment]
@mminer you should have most of what you need to dynamically load parent and child lists in v0.6.0-0. See the example at: https://codesandbox.io/s/react-command-palette-dynamic-lists-p2xo9

I'll working on a few UX improvement soon

**Comment by mminer** [comment]
Thanks for looking into this. One problem that prevents `onChange` from solving this problem is that the input value resets whenever the `commands` prop changes, making it impossible to change the commands on every keystroke: https://github.com/asabaylus/react-command-palette/blob/master/src/command-palette.js#L125-L131

Without knowing the design considerations that went into this behaviour, could it simply be removed? Alternatively, I implemented the solution I described earlier of accepting a callback as the `commands` prop, which sidesteps this problem: https://github.com/StreakYC/react-command-palette/commit/766e03b531964b93433dac8c81e677f98a735e8e

**Comment by asabaylus** [comment]
@mminer youre right, I’m working on adding a defaultInputValue and some associated changes that will fix this issue.

**Comment by asabaylus** [comment]
@mminer I've just push out a new _defaultInputValue_ prop that the should enable you to setup the sort of dynamic lists you had in mind. I'm using this to develop the VS Code theme which needs to support dynamic command lists. Here's an example:

https://www.dropbox.com/s/b7hiyv638xu8kx9/2020-01-07%2020.29.15.gif?dl=1

Next, I'm planning to prevent the input value from changing when the user navigates the list of commands, but you may have enough to get unstuck now.

**Comment by mminer** [comment]
Thanks for the changes. Version 0.7.0 gets me *close* to being able to dynamically generate commands based on the input. The only obstacle is that `onChange` fires when selecting an item, making it impossible to differentiate between what the user has typed and what they've navigated to.

I guess this is what you're alluding to when you say that preventing the input value from changing breaks accessibility?

For reference, here's a barebones component I'm working with that generates a dynamic option on the fly. Selecting the dynamic option causes `onChange` to fire, which changes the dynamic option when that's not wanted.

```javascript
import React, {useState} from 'react';
import CommandPalette from 'react-command-palette';

export default function MyCommandPalette() {
  const [inputValue, setInputValue] = useState('');

  const commands = [
    {name: 'Static option', command() {}},
    {name: 'Dynamic option: ' + inputValue, command() {}}
  ];

  return <CommandPalette commands={commands} onChange={setInputValue} />;
}
```

**Comment by asabaylus** [comment]
@mminer I could add an event based upon [onSuggestionHighlighted](https://github.com/moroshko/react-autosuggest#on-suggestion-highlighted-prop) which seems to fit the criteria you’re looking for. Can you confirm this is what you need?

**Comment by mminer** [comment]
@asabaylus: Hmm, maybe? What I'm trying to get is what the user has typed, and *only* what they've typed. I'm not totally sure how I could use `onSuggestionHighlighted` alongside `onChange` to get this value, but very open to suggestions.

**Comment by asabaylus** [comment]
I see, I misunderstood. Would returning the user's query in the onChange or onSelect payload meet your needs? If not it could be added to the command callback as you'd originally proposed

**Comment by mminer** [comment]
Returning the user's query in `onChange` should do the job.

**Comment by asabaylus** [comment]
Ok, if you’d like to make a PR I’ll happily take it, otherwise I’ll probably get to this over the weekend.

**Comment by asabaylus** [comment]
@mminer I double checked, onChange returns the current input field value. 
https://www.dropbox.com/s/rt1btkb3dcopy3i/2020-01-11%2015.56.58.gif?dl=0

Is that what you need?

**Comment by mminer** [comment]
`onChange` returns the current input field value, but this value changes in two ways:

- When the user types a query
- When the user selects an item

To generate commands dynamically, I need to differentiate between the two cases.

![Screen Recording 2020-01-13 at 10 51 20 2020-01-13 10_52_31](https://user-images.githubusercontent.com/267284/72283206-24dcda00-35f3-11ea-9e7a-e4982bfda06d.gif)

**Comment by asabaylus** [comment]
@mminer onSelect will have the selected value only but fires only when the user makes their selection. Or did you need both in the onChange?

**Comment by mminer** [comment]
What would be ideal is an `onChange` function that *doesn't* fire when the user navigates between items. I only want to know what the user has explicitly typed themselves.

**Comment by asabaylus** [comment]
@mminer I added the 'query' we discussed. The query ignore up/down arrow key navigation, selection and mouse highlights. Would you please review PR #453?

Note that you'll likely need to nvm install as I had to bump node to 13.6.0.

**Comment by asabaylus** [comment]
Added a userQuery arg to onChange which resolves this issue in v0.9.9-0
